javascriptlib/src/mill/javascriptlib/PublishModule.scala

```diff
@@ -0,0 +1,539 @@
+  }
+
+  // Define exports for the package
+  // by default: mainFile is exported
+  // use this to define other exports
+  def exports: T[Map[String, String]] = Task { Map.empty[String, String] }
+
+  private def mkExports: T[Map[String, PublishModule.ExportEntry]] = Task {
+    exports().map { case (key, value) =>
+      key -> PublishModule.Export("./" + bundledOut() + "/" + value)
+    }
+  }
+
+  private def mkTypesVersion: T[Map[String, Seq[String]]] = Task {
+    pubAllSources().map { source =>
+      val dist = source.replaceFirst("typescript", bundledOut())
+      val declarations = source.replaceFirst("typescript", declarationOut())
+      ("./" + dist).replaceAll("\\.ts", "") -> Seq(declarations.replaceAll("\\.ts", ".d.ts"))
+    }.toMap
+  }
+
+  // build package.json from publishMeta
+  // mv to publishDir.dest
+  def packageJson: T[Unit] = Task { // PathRef
+    def splitDeps(input: String): (String, String) = {
+      input.split("@", 3).toList match {
+        case first :: second :: tail if input.startsWith("@") =>
+          ("@" + first + "@" + second, tail.mkString)
+        case first :: tail =>
+          (first, tail.mkString)
+      }
+    }
+
+    val json = publishMeta()
+    val updatedJson = json.copy(
+      files = json.files ++ Seq(bundledOut(), declarationOut()),
+      main = main(),
+      types = mainType(),
+      exports = Map("." -> PublishModule.Export("./" + main())) ++ mkExports(),
+      bin = json.bin.map { case (k, v) => (k, "./" + bundledOut() + "/" + v) },
+      typesVersions = mkTypesVersion(),
+      dependencies = transitiveNpmDeps().map { deps => splitDeps(deps) }.toMap,
+      devDependencies = transitiveNpmDeps().map { deps => splitDeps(deps) }.toMap
+    ).toJsonClean
+
+    os.write.over(publishDir().path / "package.json", updatedJson)
+  }
+
+  // Package.Json construction
+
+  // Compilation Options
+  override def modulePaths: Task[Seq[(String, String)]] = Task.Anon {
+    val module = millSourcePath.last
+
+    Seq((s"$module/*", "typescript/src" + ":" + s"${declarationOut()}")) ++
+      resources().map { rp =>
+        val resourceRoot = rp.path.last
+        val result = (
+          s"@$module/$resourceRoot/*",
+          resourceRoot match {
+            case s if s.contains(".dest") => rp.path.toString
+            case _ => s"typescript/$resourceRoot"
+          }
+        )
+        result
+      }
+  }
+
+  private def pubModDeps: T[Seq[String]] = Task {
+    moduleDeps.map { x =>
+      x.millSourcePath.subRelativeTo(Task.workspace).toString.split("/").head
+    }.distinct
+  }
+
+  private def pubModDepsSources: T[Seq[PathRef]] = Task {
+    for {
+      modSource <- Task.traverse(moduleDeps)(_.sources)
+    } yield modSource
+  }
+
+  private def pubModDepsGenSources: T[Seq[PathRef]] = Task {
+    (for {
+      modSource <- Task.traverse(moduleDeps)(_.generatedSources)()
+    } yield {
+      val fileExt: Path => Boolean = _.ext == "ts"
```

Comment:
Just inline these, the indirection they create just makes the code more confusing